Add types for 'mg-api-js'

kaden-sharpin opened the pull request:
Please fill in this template.
- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `pnpm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If adding a new definition:
- [x] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [x] If this is for an npm package, match the name. If not, do not conflict with the name of an npm package.
- [x] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [x] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [x] `tsconfig.json` [should have](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#tsconfigjson) `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

peterblazejewicz commented on types/mg-api-js/index.d.ts:
```diff
@@ -0,0 +1,17735 @@
+             */
+            groups: unknown[];
+            /**
+             * The unique hardware identifier for the Device. This is a 4 byte unique number to identify devic
```
```suggestion
export as namespace GeotabApi;
```
https://www.typescriptlang.org/docs/handbook/declaration-files/templates/module-d-ts.html#optional-global-usage
(using built-in, documented syntax, outcome is same)
were this generated?

peterblazejewicz replied:
wow, GitHub lost track of the code snip :)
Skip this one, just was that code generated question stands

DangerBotOSS commented:
<!--
  0 failure: 
  0 warning: 
  
  1 markdown notices
  DangerID: danger-id-Danger;
-->



## Formatting

The following files are not formatted:
1. types/mg-api-js/index.d.ts

Consider running `pnpm dprint fmt` on these files to make review easier.
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 9171bb895e47f46346e94401b25de1fa3f250282
</p>

kaden-sharpin commented:
> What was used to generate this? Is the same tool always going to generate it? Should we exclude this from formatting checks? I'm not seeing any info about how these types were created.

I wrote a script to pull the types from the [Geotab documentation](https://geotab.github.io/sdk/software/api/reference/) probably around 6 months ago. I just ran this in the browser console. It isn't pretty but it worked and I've been using these types since then. 

```javascript
// @ts-check
//
// https://geotab.github.io/sdk/software/api/reference/
//

/**
 * @param {string} selector
 * @param {Element?} [element]
 * @returns {HTMLElement | null}
 */
function $(selector, element) {
  return (element ?? document).querySelector(selector);
}

/**
 * @param {string} selector
 * @param {Element?} [element]
 * @returns {HTMLElement[]}
 */
function $$(selector, element) {
  // @ts-ignore
  return [...(element ?? document).querySelectorAll(selector)];
}

/**
 * @param {string} expression
 * @param {Element?} [element]
 */
function xpath(expression, element) {
  const result = document.evaluate(expression, element ?? document);
  /** @type {Iterable<Node>} */
  const iterable = {
    [Symbol.iterator]: function* () {
      let el;
      while ((el = result.iterateNext())) {
        yield el;
      }
    },
  };
  return [...iterable];
}

let output = "";

const methods = $$("#dataStoreMembers .member h3")
  .map((el) => JSON.stringify(el.innerText.replace(/^(\w+)(.|\n)*/, "$1")))
  .join("\n  | ");
const parameters = $$("#dataStoreMembers .member")
  .map((el) => {
    const method = JSON.stringify(
      $("h3", el)?.innerText.replace(/^(\w+)(.|\n)*/, "$1")
    );
    const description = $("p.member-description", el)?.innerText ?? "";
    const returnDesc = $("dl.member-parameters-list", el)?.innerText ?? "";
    const comment =
      !description && !returnDesc
        ? ""
        : `  /**\n   * ${description.replace(
            /\n/g,
            "\n   * "
          )}${returnDesc.replace(/\n|(^(?=\w))/g, "\n   * ")}\n   */\n`;
    /** @ts-ignore @type {HTMLTableElement | null} */
    const table = $(".member-parameters .member-parameters-list", el);

    if (!table) return comment + `${method}: {}`;

    return (
      comment +
      `  ${method}: {\n${[...table.rows]
        .map((el) => {
          let parameter = el.cells[0].innerText;
          if (parameter.endsWith(" *")) {
            parameter = JSON.stringify(parameter.slice(0, -2));
          } else {
            if (parameter.endsWith(" Beta")) parameter = parameter.slice(0, -5);
            parameter = JSON.stringify(parameter) + "?";
          }

          let type = el.cells[1].innerText.trim();
          if (type === "array") type = "Array<unknown>";

          return (
            `    /**${el.cells[2].innerText.replace(
              /\n|(^(?=\w))/g,
              "\n     * "
            )}\n     */\n` + `    ${parameter}: ${type};`
          );
        })
        .join("\n")}\n  }`
    );
  })
  .join(";\n");

const objects = $$("#objectMembers > .member")
  .map((el) => {
    const name = $("h3.member-name", el)?.innerText.replace(
      /^(\w+)(.|\n)*/,
      "$1"
    );
    const description = xpath("./text() | ./p | ./ul", $(".member-info", el))
      .reduce(
        /** @param {string[]} p */ (p, c) => {
          if (c instanceof HTMLElement) p.push(c.innerText);
          else {
            const str = c.textContent ?? "";
            p.push(...str.split("\n").map((el) => el.trim()));
          }
          return p;
        },
        []
      )
      .filter((el) => el)
      .join("\n");
    const comment = !description
      ? ""
      : `  /**\n   * ${description.replace(/\n/g, "\n   * ")}\n   */\n`;

    /** @ts-ignore @type {HTMLTableSectionElement} */
    const table = $("table.member-parameters-list > tbody", el);
    const fields = [...(table?.rows ?? [])].map((row) => {
      const name = row.cells[0].innerText;
      let type = row.cells[1].innerText.trim() || "unknown";
      const desc = row.cells[2].innerText
        .split("\n")
        .map((el) => el.trim())
        .filter((el) => el)
        .join("\n");
      const comment = !desc
        ? ""
        : `    /**\n     * ${desc.replace(/\n/g, "\n     * ")}\n     */\n`;

      if (type === "array") type = "Array<unknown>";

      return comment + `    ${JSON.stringify(name)}: ${type};`;
    });

    if (!fields.length) return comment + `  export type ${name} = {};`;
    return comment + `  export type ${name} = {\n${fields.join("\n")}\n  };`;
  })
  .join("\n");

const allObjects = $$("#objectMembers > .member").map((el) => {
  const name = $("h3.member-name", el)?.innerText.replace(
    /^(\w+)(.|\n)*/,
    "$1"
  );
  return `GeotabApiObjects.${name}`;
});

output += "type GeotabApiMethod = \n  | ";
output += methods;
output += "\n\n";

output += "type GeotabApiParameters = {\n";
output += parameters;
output += "\n}\n\n";

output += "namespace GeotabApiObjects {\n";
output += objects;
output += "\n}\n\n";

output += "type GeotabApiObject = ";
output += "\n  | " + allObjects.join("\n  | ");
output += ";\n\n";

console.log(output);
```

kaden-sharpin commented:
Ready to merge
